js-peer/src/lib/direct-message.ts

```diff
@@ -0,0 +1,180 @@
+  NO_METADATA: 'No metadata in response',
+  STATUS_NOT_OK: (status: dm.Status) => `Received status: ${status}, expected OK`
+}
+
+export interface DirectMessageEvent {
+  content: string
+  type: string
+  stream: Stream
+  connection: Connection
+}
+
+export interface DirectMessageEvents {
+  'message': CustomEvent<DirectMessageEvent>
+}
+
+interface DirectMessageComponents {
+  registrar: Registrar
+  connectionManager: ConnectionManager
+}
+
+export class DirectMessage extends TypedEventEmitter<DirectMessageEvents> implements Startable {
+  readonly [serviceDependencies]: string[] = [
+    '@libp2p/identify',
+    '@libp2p/connection-encryption',
+    '@libp2p/transport',
+    '@libp2p/stream-multiplexing',
+  ]
+
+  readonly [serviceCapabilities]: string[] = [
+    '@universal-connectivity/direct-message'
+  ]
+
+  private topologyId?: string
+  private readonly components: DirectMessageComponents
+  private dmPeers: Set<string> = new Set()
+
+  constructor(components: DirectMessageComponents) {
+    super()
+    this.components = components
+  }
+
+  async start(): Promise<void> {
+    this.topologyId = await this.components.registrar.register(DIRECT_MESSAGE_PROTOCOL, {
+      onConnect: this.handleConnect.bind(this),
+      onDisconnect: this.handleDisconnect.bind(this)
+    })
+  }
+
+  async afterStart(): Promise<void> {
+    await this.components.registrar.handle(
+      DIRECT_MESSAGE_PROTOCOL,
+      async ({ stream, connection }) => {
+        await this.receive(stream, connection)
+      },
+    )
+  }
+
+  stop(): void {
+    if (this.topologyId != null) {
+      this.components.registrar.unregister(this.topologyId)
+    }
+  }
+
+  private handleConnect(peerId: PeerId): void {
+    this.dmPeers.add(peerId.toString())
+  }
+
+  private handleDisconnect(peerId: PeerId): void {
+    this.dmPeers.delete(peerId.toString())
+  }
+
+  isDMPeer(peerId: PeerId): boolean {
+    return this.dmPeers.has(peerId.toString())
+  }
+
+  async send(peerId: PeerId, message: string): Promise<boolean> {
+    if (!message) {
+      throw new Error(ERRORS.EMPTY_MESSAGE)
+    }
+
+    let stream: Stream | undefined
+
+    try {
+      const conn = await this.components.connectionManager.openConnection(peerId, { signal: AbortSignal.timeout(5000) })
+      if (!conn) {
+        throw new Error(ERRORS.NO_CONNECTION)
+      }
+
+      // Single protocols can skip full negotiation
+      const stream = await conn.newStream(DIRECT_MESSAGE_PROTOCOL, { negotiateFully: false })
+
+      if (!stream) {
+        throw new Error(ERRORS.NO_STREAM)
+      }
+
+      const datastream = pbStream(stream)
+
+      const req: dm.DirectMessageRequest = {
+        content: message,
+        type: MIME_TEXT_PLAIN,
+        metadata: {
+          clientVersion: dmClientVersion,
+          timestamp: BigInt(Date.now()),
+        },
+      }
+
+      await datastream.write(req, dm.DirectMessageRequest)
+
+      const res = await datastream.read(dm.DirectMessageResponse)
```

Comment:
It's a good idea to pass a signal here so you don't wait forever:

```suggestion
      await datastream.write(req, dm.DirectMessageRequest, {
        signal: AbortSignal.timeout(5000)
      })

      const res = await datastream.read(dm.DirectMessageResponse, {
        signal: AbortSignal.timeout(5000)
      })
```

The above has two separate 5s timeouts, so it'll take a max 10s for the write/read operation - you may wish to shorten this or reuse the same signal object in both operations.

Reply:
Makes sense - thank you!